### ISSUE TITLE
A question about container monitoring in kata vm

### ISSUE THREAD

**Issue by singhwang** [kata-containers/kata-containers#6179]
I have tried to obtain cgroups information in kata vm through CRI API and ShimV2API respectively, but there is no blkio and network IO related information in this part of information, mainly blkio, I need to use this information to construct container disk IO related monitoring indicators . How to solve this problem?

**Comment by liubin** [comment]
I tested and found there is an issue in cgroup-rs to handle `Discard` item in the `blkio.throttle.io_service_bytes`:

```
254:0 Read 695296
254:0 Write 4096
254:0 Sync 699392
254:0 Async 0
254:0 Discard 0
254:0 Total 699392
Total 699392
```

I created an [issue](https://github.com/kata-containers/cgroups-rs/issues/109) to fix it.

**Comment by singhwang** [comment]
@liubin Thank you very much, Teacher Liu Bin! You always answer my questions. Can you post your test command? I want to learn, I actually test blkio cgroups have no value at all.

**Comment by singhwang** [comment]
@liubin In addition, the question you mentioned is "Note that virtiofs has no blkio stats, you may need some block devices." I want to ask again, the shared_fs of kata can only use virtio-9p, virtio-fs or virtio-fs-nydus , I don't quite understand what you said about using block devices here.

**Comment by liubin** [comment]
If you attach a block device from host to container(guest), then you will get blkio stats.

**Comment by singhwang** [comment]
> If you attach a block device from host to container(guest), then you will get blkio stats.

Got it, thanks for your answer! I also want to ask, if I use kata to start a pod, I want to get the monitoring data of the disk IO of each container under the pod, how should I do it? I understand that the rootfs of the container is on the host machine at this time, and is shared in the vm through virtiofs.

**Comment by Ji-Xinyou** [comment]
> Are you referring to changing containerd's storage driver (snapshotter) from overlayfs to device mapper, and using block devices to provide container rootfs?

IMO using device mapper to create a block devices, and use that to provide container rootfs could be a solution, since it is not using virtiofs anymore.

**Comment by singhwang** [comment]
@Ji-Xinyou  Thanks for your answer. What is it IMO? I do not understand. Using virtiofs to share the rootfs to the pod vm, is there still a way to monitor the disk IO of the container in the pod vm?

**Comment by Ji-Xinyou** [comment]
> @Ji-Xinyou Thanks for your answer. What is it IMO? I do not understand. Using virtiofs to share the rootfs to the pod vm, is there still a way to monitor the disk IO of the container in the pod vm?

IMO stands for In My Opinion :D

I think using virtiofs cannot generate blockio stats. Currently I can only think of using a block device and plug it into the VM.

**Comment by singhwang** [comment]
I used the block device created by device mapper to provide the rootfs of the container, then used fio tool to perform read and write tests, and entered the pod vm through the debug console to view the blkio cgroups information, but could not read any data. Does anyone know what's going on?

Use the fio tool to do read and write tests separately:
```
$ fio -ioengine=psync -bs=4k -direct=1 -numjobs=1 -thread -fallocate=none -rw=read -filename=/mnt/orangefs/fio.4k.read -name="BS 4KB read test" -runtime=1 -size=10M

$ fio -ioengine=psync -bs=4k -direct=1 -numjobs=1 -thread -fallocate=none -rw=write -filename=/mnt/orangefs/fio.4k.write -name="BS 4KB read write" -runtime=1 -size=10M
```

Use the debug console to enter the pod vm to view blkio cgroups information:
```
$ cd /sys/fs/cgroup/blkio/cri-containerd/cb9509bcac1ca72ee04c960ed7380bc95447df5146b58227034b7b644aaa477c/
$ cat blkio.throttle.io_serviced
8:16 Read 0
8:16 Write 0
8:16 Sync 0
8:16 Async 0
8:16 Discard 0
8:16 Total 0
Total 0

$ cat blkio.throttle.io_service_bytes
8:16 Read 0
8:16 Write 0
8:16 Sync 0
8:16 Async 0
8:16 Discard 0
8:16 Total 0
Total 0
```

**Comment by liubin** [comment]
@singhwang how did you insert the block device? Can you share your config files?

**Comment by singhwang** [comment]
@liubin  I just set containerd's storage driver (snapshotter) from overlayfs to device mapper, and using block devices to provide container rootfs. 
```
$ cat /etc/containerd/config.toml
......
    [plugins."io.containerd.grpc.v1.cri".containerd]
      default_runtime_name = "runc"
      disable_snapshot_annotations = true
      discard_unpacked_layers = false
      no_pivot = false
      snapshotter = "devmapper"
......
  [plugins."io.containerd.snapshotter.v1.devmapper"]
    async_remove = false
    pool_name = "devpool"
    root_path = "/var/lib/containerd/devmapper"
    base_image_size = "2GB"
    discard_blocks = true
......
```
In this way, kata will use the block_device_driver in the configuration file (default is /opt/kata/share/defaults/kata-containers/configuration-qemu.toml), the default is virtio-scsi, and the rootfs of the container will be sda in the pod vm , sdb appears in this form.

**Comment by liubin** [comment]
I created a Pod that mount a host device(`/dev/mapper/loop12p1`) into container:


```yaml

apiVersion: v1
kind: Pod
metadata:
  name: kata
spec:
  runtimeClassName: kata
  containers:
    - name: sbt
      image: docker.io/containerstack/alpine-stress
      imagePullPolicy: Always
      command: ["sh", "-c"]
      args:
        - tail -f /dev/null
      securityContext:
        privileged: true
      volumeMounts:
      - mountPath: /dev/loop12p1
        name: test-volume
  volumes:
  - name: test-volume
    hostPath:
      path: /dev/mapper/loop12p1

```

And mount the device inside the container:

```bash
# kubectl exec -it kata sh
/ # mount /dev/loop12p1 /mnt
/ # dd if=/dev/zero of=/mnt/test.disk bs=1M count=10
10+0 records in
10+0 records out
```

And then can get the stats.

**Comment by singhwang** [comment]
@liubin Thanks for your help. I will learn.This is what I really want to know about what you posted. Does this situation have nothing to do with the storage driver settings of containerd? Whether my containerd storage driver is overlayfs or device mapper, can it generate disk IO on the block device?

**Comment by singhwang** [comment]
@liubin I tried the way you suggested, but I still can't read blkio cgroups info. For convenience, I did not use the virtual block device of device mapper. I used a bare disk, partitioned and formatted it, and then applied the yaml you gave. After entering the container, I also followed your prompts, and they all succeeded. The only difference is that when you enter the pod vm to see the corresponding blk cgroups information, they are still 0.

I also found that the latest pod vm is full of cgroups v2. In order to do this experiment, I deliberately used ubuntu 20.04 to recreate the rootfs image according to the official document and the corresponding tool, so that the debug console can enter to see the information of cgroups v1. Otherwise, all you see when entering the debug console is cgroups v2.

**Comment by liubin** [comment]
@singhwang do you mean that block metrics in cgroup v2 is not work?

**Comment by singhwang** [comment]
@liubin No, I haven't tested whether cgroups v2 works, and I don't know whether kata agent supports cgroups v2. However, I read a bit of containerd-shim-kata-v2 code, and found that the information returned to containerd through ShimV2API is all cgroups v1 information, and the blkio cgroups information you sent is also cgroups v1 information, so I use support The rootfs image of cgroups v1 to test.

kata-containers/src/runtime/pkg/containerd-shim-v2/metrics.go
```
func marshalMetrics(ctx context.Context, s *service, containerID string) (*google_protobuf.Any, error) {
	stats, err := s.sandbox.StatsContainer(ctx, containerID)
	if err != nil {
		return nil, err
	}

	metrics := statsToMetrics(&stats)

	data, err := typeurl.MarshalAny(metrics)
	if err != nil {
		return nil, err
	}

	return data, nil
}

func statsToMetrics(stats *vc.ContainerStats) *cgroupsv1.Metrics {
	metrics := &cgroupsv1.Metrics{}

	if stats.CgroupStats != nil {
		metrics = &cgroupsv1.Metrics{
			Hugetlb: setHugetlbStats(stats.CgroupStats.HugetlbStats),
			Pids:    setPidsStats(stats.CgroupStats.PidsStats),
			CPU:     setCPUStats(stats.CgroupStats.CPUStats),
			Memory:  setMemoryStats(stats.CgroupStats.MemoryStats),
			Blkio:   setBlkioStats(stats.CgroupStats.BlkioStats),
		}
	}

	metrics.Network = setNetworkStats(stats.NetworkStats)

	return metrics
}
```

**Comment by singhwang** [comment]
@liubin I use a bare disk to test, and there should be no difference from the virtual block device generated by your device mapper. I understand that the yaml type /dev/mapper/loop12p1 you sent should be a virtual block device that has been partitioned and formatted. Otherwise, even if it is mapped to a container, it cannot be mounted directly. Is my understanding correct?
